File: ai-service/app/schemas.py
from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, HttpUrl, field_validator


SuspicionLevel = Literal["low", "moderate", "high"]


class ModelInfo(BaseModel):
    name: str = Field(default="oral-lesion-triage-cnn")
    version: str = Field(default="1.0.0")
    architecture: str = Field(default="mobilenetv3-small")


class OralLesionInferenceRequest(BaseModel):
    request_id: str | None = None
    case_id: UUID | None = None
    case_code: str | None = None
    image_id: UUID | str
    image_url: HttpUrl
    model: ModelInfo | None = None
    model_name: str | None = None
    model_version: str | None = None

    @field_validator("case_code")
    @classmethod
    def validate_case_code(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("case_code no puede estar vacio.")
        return value

    def resolved_model_name(self, default: str) -> str:
        if self.model and self.model.name:
            return self.model.name
        return self.model_name or default

    def resolved_model_version(self, default: str) -> str:
        if self.model and self.model.version:
            return self.model.version
        return self.model_version or default

    def resolved_model_architecture(self, default: str) -> str:
        if self.model and self.model.architecture:
            return self.model.architecture
        return default


class GradcamResponse(BaseModel):
    content_type: str
    base64: str


class OralLesionInferenceResponse(BaseModel):
    service_request_id: str
    request_id: str | None = None
    model_name: str
    model_version: str
    architecture: str
    input_shape: list[int]
    suspicion_level: SuspicionLevel
    probability: float
    class_probabilities: dict[str, float]
    gradcam_base64: str | None = None
    gradcam_mime_type: str | None = None
    gradcam: GradcamResponse | None = None
    latency_ms: int
    warnings: list[str] = Field(default_factory=list)

File: apps/web/src/features/inference/schemas.ts
import { z } from "zod"

export const suspicionLevelSchema = z.enum(["low", "moderate", "high"])
export const urgencyLevelSchema = z.enum(["routine", "priority", "urgent"])

export const runInferenceResponseSchema = z.object({
  success: z.literal(true),
  request_id: z.string(),
  data: z.object({
    case_id: z.string(),
    image_id: z.string(),
    inference_id: z.string(),
    model: z.object({
      name: z.string(),
      version: z.string(),
      architecture: z.string(),
    }),
    prediction: z.object({
      suspicion_level: suspicionLevelSchema,
      probability: z.number(),
      class_probabilities: z.record(z.string(), z.number()),
    }),
    recommendation: z.object({
      suspicion_level: suspicionLevelSchema,
      urgency_level: urgencyLevelSchema,
      professional_referral: z.boolean(),
      reason_codes: z.array(z.string()).default([]),
      message: z.string(),
    }),
    next_step: z.literal("generate_report"),
  }),
  message: z.string(),
})
